1. # Problem Statement
### Heart Disease Prediction Using Machine Learning

Cardiovascular diseases are among the leading causes of death worldwide. Early detection of heart disease can significantly improve treatment outcomes and reduce mortality. However, diagnosing heart disease often requires multiple clinical examinations and expert medical analysis.

Machine learning can assist healthcare professionals by analyzing patient health records and identifying patterns associated with heart disease. The goal of this project is to develop a predictive model that estimates whether a patient is likely to have heart disease based on various medical attributes.

### Important Note:
This project is intended to assist healthcare professionals by providing decision support. It is not designed to replace medical diagnosis or professional clinical judgment.

2. # Project Objective

The primary objective of this project is to build and evaluate multiple machine learning models capable of predicting the presence of heart disease using patient medical information.

The project aims to:

- Understand the dataset and its medical features.
- Perform exploratory data analysis (EDA) to identify trends and relationships.
- Clean and preprocess the dataset.
- Compare multiple machine learning algorithms.
- Optimize the best-performing model using hyperparameter tuning.
- Evaluate model performance using appropriate classification metrics.
- Explain model predictions using Explainable AI (SHAP).
- Deploy the final model through a web application or REST API.

In [ ]:
import pandas as pd

heart_data = pd.read_csv('../data/raw/heart_disease_data.csv')

In [11]:
heart_data.shape

(303, 14)

In [12]:
heart_data.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1


In [13]:
heart_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


3. # Dataset Information
### Dataset Name

Heart Disease Dataset (Youtuber Siddhardhan)

### Dataset Source

The dataset is provided to us from the Youtuber Siddhardhan and he might have fetched from UCI or Kaggle

### Dataset Characteristics
| Property           | Description                             |
| ------------------ | --------------------------------------- |
| Problem Type       | Binary Classification                   |
| Target Variable    | target                                  |
| Number of Features | 13                                      |
| Target Classes     | 0 = No Heart Disease, 1 = Heart Disease |
| Data Type          | Structured Tabular Data                 |

### Target Variable

The target column represents whether a patient has heart disease.
| Value | Meaning               |
| ----- | --------------------- |
| 0     | No Heart Disease      |
| 1     | Heart Disease Present |


4. # Feature Description
| Feature  | Description                                        | Data Type            |
| -------- | -------------------------------------------------- | -------------------- |
| age      | Age of the patient in years                        | Numerical            |
| sex      | Gender (1 = Male, 0 = Female)                      | Binary               |
| cp       | Chest pain type                                    | Categorical          |
| trestbps | Resting blood pressure (mm Hg)                     | Numerical            |
| chol     | Serum cholesterol (mg/dl)                          | Numerical            |
| fbs      | Fasting blood sugar >120 mg/dl                     | Binary               |
| restecg  | Resting electrocardiographic results               | Categorical          |
| thalach  | Maximum heart rate achieved                        | Numerical            |
| exang    | Exercise-induced angina                            | Binary               |
| oldpeak  | ST depression induced by exercise relative to rest | Numerical            |
| slope    | Slope of the peak exercise ST segment              | Categorical          |
| ca       | Number of major vessels colored by fluoroscopy     | Numerical (Discrete) |
| thal     | Thalassemia status                                 | Categorical          |
| target   | Presence of heart disease                          | Target Variable      |

5. # Feature Classification
### Numerical Features
- age
- trestbps
- chol
- thalach
- oldpeak
- ca
### Categorical Features
- cp
- restecg
- slope
- thal
### Binary Features
- sex
- fbs
- exang
### Target Variable
- target

6. # Initial Assumptions

Before performing any analysis, the following assumptions are made:

### Data Quality

The dataset may contain:

- Missing values
- Duplicate records
- Outliers
- Incorrect data types

These issues will be verified during Exploratory Data Analysis.

### Feature Scaling

Some algorithms such as:

- Logistic Regression
- K-Nearest Neighbors (KNN)
- Support Vector Machine (SVM)

are sensitive to feature magnitude. Therefore, numerical features will likely require feature scaling using either StandardScaler or MinMaxScaler.

### Class Distribution

The distribution of the target classes is currently unknown. During EDA, we will determine whether the dataset is balanced or imbalanced.

If the dataset is significantly imbalanced, relying solely on accuracy may produce misleading results.

### Feature Relationships

Certain medical attributes, such as chest pain type, maximum heart rate, and exercise-induced angina, are expected to have a strong relationship with the target variable.

Correlation analysis and feature importance techniques will be used to validate these assumptions.

### Model Selection

Several machine learning algorithms will be compared before selecting the final model. The project will not assume that Logistic Regression is the best solution.

Candidate algorithms include:

- Logistic Regression
- K-Nearest Neighbors
- Support Vector Machine
- Decision Tree
- Random Forest
- XGBoost

7. # Evaluation Metric Discussion

Since this is a binary classification problem, evaluating the model using only accuracy may not provide a complete picture of performance.

For example, consider a dataset where:

- 90% of patients do not have heart disease.
- 10% of patients do have heart disease.

A model that predicts every patient as healthy would achieve 90% accuracy, despite failing to identify a single patient with heart disease. Such a model would be clinically useless.

Therefore, additional evaluation metrics will be considered.

### Accuracy

Measures the proportion of correctly classified instances.

Useful when class distributions are balanced but insufficient on its own for medical datasets.

### Precision

Precision measures how many patients predicted as having heart disease actually have the disease.

High precision reduces false alarms.

### Recall (Sensitivity)

Recall measures how many patients with heart disease are correctly identified by the model.

In medical diagnosis, recall is often more important than accuracy because failing to detect a patient with heart disease (False Negative) can delay treatment and increase health risks.

### F1 Score

The F1 Score provides a balance between precision and recall, making it useful when both false positives and false negatives are important.

### ROC-AUC Score

The ROC-AUC Score evaluates the model's ability to distinguish between patients with and without heart disease across different classification thresholds.

A higher ROC-AUC indicates better overall discrimination capability.

### Confusion Matrix

The confusion matrix provides a detailed summary of prediction outcomes:

- True Positives (TP)
- True Negatives (TN)
- False Positives (FP)
- False Negatives (FN)

This helps identify the types of errors the model makes.

# Conclusion

At this stage, the problem has been clearly defined, the dataset has been introduced, the features have been understood, and appropriate evaluation metrics have been selected.

The next phase will focus on Exploratory Data Analysis (EDA) to investigate the dataset, identify data quality issues, discover relationships between variables, and prepare the data for machine learning.